File: src/web/schemas/sentiment.py
"""Pydantic models for sentiment analysis and cross-market correlation.

Per PRD v3.2 FR-TN003 (resonance), FR-TN004 (sentiment report),
FR-GM004 (cross-market).
"""

from pydantic import BaseModel


# --- FR-TN003: Resonance ---


class ResonanceTimelineEntry(BaseModel):
    platform: str
    rank: int
    timestamp: str


class ResonanceEventItem(BaseModel):
    event_id: str
    title: str
    resonance_level: str
    platforms: list[str]
    rank_timeline: list[ResonanceTimelineEntry] = []
    related_stocks: list[str] = []
    sentiment: str = "neutral"
    first_appeared: str = ""
    last_updated: str = ""
    heat_score: float = 0.0


class ResonanceResult(BaseModel):
    status: str
    events: list[ResonanceEventItem]
    total: int
    generated_at: str


# --- FR-TN004: Sentiment Report ---


class SentimentCoreTrend(BaseModel):
    topic: str
    resonance_level: str = "L1"
    sentiment: str = "neutral"
    related_stocks: list[str] = []
    summary: str = ""


class SentimentPolicySignal(BaseModel):
    title: str
    impact: str = "neutral"
    affected_sectors: list[str] = []
    confidence: float = 0.5
    summary: str = ""


class SentimentGlobalLinkage(BaseModel):
    us_market_summary: str = ""
    commodity_impact: str = ""
    forex_impact: str = ""


class SentimentRiskAlert(BaseModel):
    type: str = ""
    title: str = ""
    severity: str = "medium"
    mitigation: str = ""


class SentimentSectorOutlook(BaseModel):
    bullish: list[str] = []
    bearish: list[str] = []
    neutral: list[str] = []


class SentimentReportResult(BaseModel):
    status: str
    core_trends: list[SentimentCoreTrend] = []
    policy_signals: list[SentimentPolicySignal] = []
    global_linkage: SentimentGlobalLinkage = SentimentGlobalLinkage()
    risk_alerts: list[SentimentRiskAlert] = []
    sector_outlook: SentimentSectorOutlook = SentimentSectorOutlook()
    overall_outlook: str = ""
    generated_at: str = ""
    disclaimer: str = ""
    model_used: str | None = None
    message: str | None = None


# --- FR-GM004: Cross-Market ---


class CrossMarketPeer(BaseModel):
    symbol: str
    price: float | None = None
    pct_change: float | None = None


class CrossMarketGroup(BaseModel):
    trend: str = "neutral"
    peers: list[CrossMarketPeer] = []
    impact_score: float = 0.0
    avg_pct_change: float | None = None
    summary: str | None = None


class CrossMarketResult(BaseModel):
    symbol: str
    tags: list[str] = []
    us_market: CrossMarketGroup = CrossMarketGroup()
    hk_market: CrossMarketGroup = CrossMarketGroup()
    commodity_exposure: CrossMarketGroup = CrossMarketGroup()
    global_indices: CrossMarketGroup = CrossMarketGroup()
    combined_impact_score: float = 0.0
    impact_direction: str = "neutral"
    generated_at: str = ""

File: frontend/src/types/sentiment.ts
/** Sentiment analysis and cross-market types — PRD v3.2 FR-TN003~005, FR-GM004 */

// FR-TN003: Resonance events
export interface ResonanceTimelineEntry {
  platform: string
  rank: number
  timestamp: string
}

export interface ResonanceEvent {
  event_id: string
  title: string
  resonance_level: string // "L1" | "L2" | "L3"
  platforms: string[]
  rank_timeline: ResonanceTimelineEntry[]
  related_stocks: string[]
  sentiment: string // "positive" | "negative" | "mixed" | "neutral"
  first_appeared: string
  last_updated: string
  heat_score: number
}

export interface ResonanceResult {
  status: string
  events: ResonanceEvent[]
  total: number
  generated_at: string
}

// FR-TN004: Sentiment report
export interface SentimentCoreTrend {
  topic: string
  resonance_level: string
  sentiment: string
  related_stocks: string[]
  summary: string
}

export interface SentimentPolicySignal {
  title: string
  impact: string
  affected_sectors: string[]
  confidence: number
  summary: string
}

export interface SentimentGlobalLinkage {
  us_market_summary: string
  commodity_impact: string
  forex_impact: string
}

export interface SentimentRiskAlert {
  type: string
  title: string
  severity: string // "low" | "medium" | "high"
  mitigation: string
}

export interface SentimentSectorOutlook {
  bullish: string[]
  bearish: string[]
  neutral: string[]
}

export interface SentimentReport {
  status: string
  core_trends: SentimentCoreTrend[]
  policy_signals: SentimentPolicySignal[]
  global_linkage: SentimentGlobalLinkage
  risk_alerts: SentimentRiskAlert[]
  sector_outlook: SentimentSectorOutlook
  overall_outlook: string
  generated_at?: string
  disclaimer?: string
  message?: string
}

// FR-TN005: Market pulse
export interface HoldingsNewsItem {
  title: string
  platform: string
  heat_score: number
  sentiment: string
}

export interface MarketPulseResult {
  status: string
  hot_events: ResonanceEvent[]
  holdings_news: Record<string, HoldingsNewsItem[]>
  global_snapshot: {
    indices: { name: string; pct_change: number | null }[]
    commodities: { name: string; pct_change: number | null }[]
    currencies: { name: string; price: number | null }[]
  } | null
  generated_at: string
}

// FR-GM004: Cross-market
export interface CrossMarketPeer {
  symbol: string
  price: number | null
  pct_change: number | null
}

export interface CrossMarketGroup {
  trend: string
  peers: CrossMarketPeer[]
  impact_score: number
  avg_pct_change?: number
  summary?: string
}

export interface CrossMarketResult {
  symbol: string
  tags: string[]
  us_market: CrossMarketGroup
  hk_market: CrossMarketGroup
  commodity_exposure: CrossMarketGroup
  global_indices: CrossMarketGroup
  combined_impact_score: number
  impact_direction: string
  generated_at: string
}
